From 5f587bef198973cf80a20979b17e60a828c5bddf Mon Sep 17 00:00:00 2001
From: Marcos Paulo de Souza <mpdesouza@suse.com>
Date: Tue, 12 Mar 2024 22:52:58 -0300
Subject: [PATCH] Add ipa-clones and -flive-patching GCC flags for livepatching

This is being used in the SUSE Linux Enterprise kernels. Add them to
easily use klp-build extract to create source-based livepatches.

Signed-off-by: Marcos Paulo de Souza <mpdesouza@suse.com>
---
 Makefile                 | 10 ++++++++++
 kernel/livepatch/Kconfig |  8 ++++++++
 2 files changed, 18 insertions(+)

diff --git a/Makefile b/Makefile
index d51d411d44a8..40f840e1e55a 100644
--- a/Makefile
+++ b/Makefile
@@ -751,6 +751,12 @@ CFLAGS_GCOV	+= -fno-tree-loop-im
 endif
 export CFLAGS_GCOV
 
+ifdef CONFIG_LIVEPATCH_IPA_CLONES
+ifeq ($(KBUILD_EXTMOD),)
+KBUILD_CFLAGS += -fdump-ipa-clones
+endif
+endif
+
 # The arch Makefiles can override CC_FLAGS_FTRACE. We may also append it later.
 ifdef CONFIG_FUNCTION_TRACER
   CC_FLAGS_FTRACE := -pg
@@ -931,6 +937,10 @@ KBUILD_RUSTFLAGS_KERNEL += -Zfunction-sections=y
 LDFLAGS_vmlinux += --gc-sections
 endif
 
+ifdef CONFIG_LIVEPATCH
+KBUILD_CFLAGS += $(call cc-option, -flive-patching=inline-clone)
+endif
+
 ifdef CONFIG_SHADOW_CALL_STACK
 ifndef CONFIG_DYNAMIC_SCS
 CC_FLAGS_SCS	:= -fsanitize=shadow-call-stack
diff --git a/kernel/livepatch/Kconfig b/kernel/livepatch/Kconfig
index 53d51ed619a3..9a6f70f1d7eb 100644
--- a/kernel/livepatch/Kconfig
+++ b/kernel/livepatch/Kconfig
@@ -18,3 +18,11 @@ config LIVEPATCH
 	  module uses the interface provided by this option to register
 	  a patch, causing calls to patched functions to be redirected
 	  to new function code contained in the patch module.
+
+config LIVEPATCH_IPA_CLONES
+       bool "Kernel Live Patching - Dump IPA Clones"
+       depends on LIVEPATCH
+       depends on $(cc-option, -fdump-ipa-clones)
+       help
+         Let GCC dump IPA clones during compilation.
+         Say N if you are unsure.
-- 
2.44.0

